src/internal.c

```diff
@@ -776,19 +794,231 @@ union wolfSSH_key {
+            }
+
+            wc_FreeRsaKey(&key->rsa);
+        }
+    }
+#endif
+#ifndef WOLFSSH_NO_ECDSA
+    if (ret == WS_SUCCESS) {
+        if (keyId == ID_UNKNOWN) {
+            scratch = 0;
+            ret = wc_ecc_init_ex(&key->ecc, heap, INVALID_DEVID);
+
+            if (ret == 0) {
+                if (isPrivate) {
+                    ret = wc_EccPrivateKeyDecode(in, &scratch,
+                            &key->ecc, inSz);
+                }
+                else {
+                    ret = wc_EccPublicKeyDecode(in, &scratch,
+                            &key->ecc, inSz);
+                }
+
+                /* If decode was successful, this was an RSA key. */
+                if (ret == 0) {
+                    switch (wc_ecc_get_curve_id(key->ecc.idx)) {
+                        case ECC_SECP256R1:
+                            keyId = ID_ECDSA_SHA2_NISTP256;
+                            break;
+                        case ECC_SECP384R1:
+                            keyId = ID_ECDSA_SHA2_NISTP384;
+                            break;
+                        case ECC_SECP521R1:
+                            keyId = ID_ECDSA_SHA2_NISTP521;
+                            break;
+                    }
+                }
+                /* in case we need to check for another key flavor */
+                ret = WS_SUCCESS;
+            }
+
+            wc_ecc_free(&key->ecc);
+        }
+    }
+#endif /* ! WOLFSSH_NO_ECDSA */
+
+    if (key != NULL) {
+        WFREE(key, heap, dynType);
+    }
+
+    if (keyId == ID_UNKNOWN) {
+        ret = WS_UNIMPLEMENTED_E;
+    }
+
+    return ret == WS_SUCCESS ? keyId : ret;
+}
+
+
+#ifdef WOLFSSH_CERTS
+/*
+ * Identifies the flavor of an X.509 certificate, RSA or ECDSA, and returns
+ * the key type ID. The process is to decode the certificate and pass the
+ * public key to IdentifyKey.
+ *
+ * @param in        certificate to identify
+ * @param inSz      size of certificate
+ * @param heap      heap to use for memory allocation
+ * @return          keyId as int, WS_MEMORY_E, WS_UNIMPLEMENTED_E
+ */
+static int IdentifyCert(const byte* in, word32 inSz, void* heap)
+{
+    struct DecodedCert cert;
+    byte *key = NULL;
+    word32 keySz = 0;
+    int ret;
+    int keyId = ID_UNKNOWN;
+
+    wc_InitDecodedCert(&cert, in, inSz, heap);
+    ret = wc_ParseCert(&cert, CERT_TYPE, 0, NULL);
+    if (ret == 0) {
+        ret = wc_GetPubKeyDerFromCert(&cert, NULL, &keySz);
+        if (ret == LENGTH_ONLY_E) {
+            ret = 0;
+            key = (byte*)WMALLOC(keySz, heap, DYNTYPE_PUBKEY);
+            if (key == NULL) {
+                ret = WS_MEMORY_E;
+            }
+        }
+    }
+
+    if (ret == 0) {
+        ret = wc_GetPubKeyDerFromCert(&cert, key, &keySz);
+    }
+
+    if (ret == 0) {
+        ret = IdentifyKey(key, keySz, 0, heap);
+        if (ret > 0) {
+            keyId = ret;
+            ret = WS_SUCCESS;
+        }
+    }
+
+    if (key != NULL) {
```

Comment:
WFREE should check for NULL, so this check can be omitted, I think.

Reply:
Check